### PR TITLE
#75 Improve the markup for displaying FormControl errors

### LINKED ISSUE #75: Improve the markup for displaying FormControl errors
### Why

Current markup for displaying proper form control error messages is complex:
```
<span *ngIf="(signupForm.get('newPasswordsGroup.confirmPassword').dirty || signupForm.get('newPasswordsGroup.confirmPassword').touched) && signupForm.get('newPasswordsGroup.confirmPassword').hasError('required')" class="app-form__element-error">
  {{ 'This field is required.' | translate }}
</span>
<span *ngIf="(signupForm.get('newPasswordsGroup.confirmPassword').dirty || signupForm.get('newPasswordsGroup.confirmPassword').touched) && signupForm.get('newPasswordsGroup.confirmPassword').hasError('passwordConfirmed')" class="app-form__element-error">
  {{ 'Passwords do not match.' | translate }}
</span>
```

### What

* Introduce a component that will deal with `touched` and `dirty` states and will display all errors for a particular control.

### PR DESCRIPTION
**Pull request by SchnWalter**
I've also added the more validators on the login ans signup page form controls. And fixed the `e2e` test and failing lint warnings.

This PR fixes #75

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -5,142 +5,147 @@
...
+        "node-sass": "^4.9.1",
+        "opn": "^5.1.0",
+        "parse5": "^4.0.0",
+        "portfinder": "^1.0.13",
+        "postcss": "^6.0.22",
+        "postcss-import": "^11.1.0",
+        "postcss-loader": "^2.1.5",
+        "postcss-url": "^7.3.2",
+        "raw-loader": "^0.5.1",
+        "rxjs": "^6.0.0",
```
Comment (cadobre) [inline package-lock.json]: why are we getting the caret symbol '^'?  rxjs": "^6.0.0" ?

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      })
+    );
+  }
+
+  public getFlatErrors(errors: ValidationErrors, parent?: string): string[] {
+    const errorPath = isEmpty(parent) ? `${this.namespace}.${this.formControlPath}.` : `${parent}.`;
+
+    const errorStrings: string[] = [];
+    Object.getOwnPropertyNames(errors).forEach((error: string) => {
+      if (isObject(errors[error])) {
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: We should add a test for this class so it's more obvious how it works.

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        return null;
+      })
+    );
+  }
+
+  public getFlatErrors(errors: ValidationErrors, parent?: string): string[] {
+    const errorPath = isEmpty(parent) ? `${this.namespace}.${this.formControlPath}.` : `${parent}.`;
+
+    const errorStrings: string[] = [];
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: errorPhrase

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    Object.getOwnPropertyNames(errors).forEach((error: string) => {
+      if (isObject(errors[error])) {
+        this.getFlatErrors(errors[error], parent).forEach((errorString: string) => {
+          errorStrings.push(errorString);
+        });
+      } else {
+        const errorString = `${errorPath}${error}`;
+        errorStrings.push(errorString);
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: errorPhraseList.push(errorPhrase);

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      map((): string[] | null => {
+        if (this.formControl.errors) {
+          return this.getFlatErrors(this.formControl.errors);
+        }
+        return null;
+      })
+    );
+  }
+
+  public getFlatErrors(errors: ValidationErrors, parent?: string): string[] {
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: getFlatErrorPhrases + add documentation

Comment (cadobre) [reply src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: who is parent? We should have a test where parent is defined.

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  @Input()
+  public readonly formControlPath: string;
+
+  @Input()
+  public readonly namespace: string;
+
+  /**
+   * A list of `ValidationErrors` that should be handled as trees.
+   */
+  protected readonly multiLevelValidationErrors: string[] = [
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: I don't see where this is used, remove.

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    'passwordPolicy',
+  ];
+
+  constructor(
+    @Host()
+    private formDir: FormGroupDirective,
+  ) {
+  }
+
+  ngOnInit() {
```
Comment (cadobre) [inline src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: I think we should explicity add public for all methods

File: package-lock.json (hunk 1 above)
Comment (SchnWalter) [reply package-lock.json]: That is what is defined by `@angular-devkit/build-angular` in it's `package.json` file. It has nothing to do with our locked version. For each of these dependencies, you have an entry in this file, with a fixed version.

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts (hunk 7 above)
Comment (SchnWalter) [reply src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: Enabled the `member-access` tslint rule.

I'll commit a fix for all classes in this PR.

File: src/app-shared/core/components/form-control-errors/form-control-errors.component.ts (hunk 2 above)
Comment (SchnWalter) [reply src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: This actually needs to be:
```ts
if (isObject(errors[errorKey]) && this.multiLevelValidationErrors.includes(errorKey)) {
  // ...
}
```
Or simply:
```ts
if (this.multiLevelValidationErrors.includes(errorKey)) {
  // ...
}
```

Comment (cadobre) [reply src/app-shared/core/components/form-control-errors/form-control-errors.component.ts]: Please also add a test

**Pull request merged**